## ECBM E4040 Homework 4 - Programming Problem

In [1]:
import theano
import hw4a
import hw4b

/Users/sanjarahmadov/anaconda/lib/python3.5/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


### Problem a: Recurrent Neural Network
#### Bullet 1: Experiment with RNNSLU

In [3]:
import hw4a

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}

hw4a.test_rnnslu(**param)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.18 best test F1 66.22
NEW BEST: epoch 1 valid F1 80.63 best test F1 77.98
NEW BEST: epoch 2 valid F1 83.79 best test F1 80.66
NEW BEST: epoch 3 valid F1 88.42 best test F1 84.92
NEW BEST: epoch 4 valid F1 89.65 best test F1 85.97
NEW BEST: epoch 5 valid F1 91.02 best test F1 86.87
NEW BEST: epoch 6 valid F1 91.45 best test F1 88.81
NEW BEST: epoch 7 valid F1 92.05 best test F1 89.89
NEW BEST: epoch 8 valid F1 93.78 best test F1 90.69
NEW BEST: epoch 9 valid F1 94.19 best test F1 91.27
NEW BEST: epoch 10 valid F1 94.7 best test F1 91.64
NEW BEST: epoch 11 valid F1 94.88 best test F1 92.23
NEW BEST: epoch 12 valid F1 95.4 best test F1 92.73
NEW BEST: epoch 13 valid F1 95.45 best test F1 92.79

NEW BEST: epoch 15 valid F1 95.55 best test F1 

(<hw4_nn.RNNSLU at 0x7fa41f739290>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

In [3]:
import hw4a

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}

exp1 = hw4a.test_rnnslu(**param)

verbose: True
normal: True
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.68 best test F1 64.92
NEW BEST: epoch 1 valid F1 79.32 best test F1 75.92
NEW BEST: epoch 2 valid F1 83.2 best test F1 80.09
NEW BEST: epoch 3 valid F1 87.1 best test F1 83.85
NEW BEST: epoch 4 valid F1 89.61 best test F1 85.38
NEW BEST: epoch 5 valid F1 90.95 best test F1 87.15
NEW BEST: epoch 6 valid F1 91.74 best test F1 88.42
NEW BEST: epoch 7 valid F1 93.77 best test F1 90.66
NEW BEST: epoch 8 valid F1 94.07 best test F1 90.74
NEW BEST: epoch 9 valid F1 94.7 best test F1 91.66
NEW BEST: epoch 10 valid F1 94.97 best test F1 91.02
NEW BEST: epoch 11 valid F1 95.54 best test F1 92.1
NEW BEST: epoch 12 valid F1 95.69 best test F1 92.39
NEW BEST: epoch 13 valid F1 95.89 best test F1 92.42
NEW BEST: epoch 14 valid F1 95.94 best test F1 92

In [5]:
import hw4a

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 70,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}

exp3 = hw4a.test_rnnslu(**param)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 70
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 70.07 best test F1 66.15
NEW BEST: epoch 1 valid F1 80.96 best test F1 77.41
NEW BEST: epoch 2 valid F1 83.99 best test F1 80.12
NEW BEST: epoch 3 valid F1 88.62 best test F1 84.49
NEW BEST: epoch 4 valid F1 90.17 best test F1 86.41
NEW BEST: epoch 5 valid F1 91.23 best test F1 87.17
NEW BEST: epoch 6 valid F1 92.41 best test F1 89.12
NEW BEST: epoch 7 valid F1 93.22 best test F1 90.43
NEW BEST: epoch 8 valid F1 94.11 best test F1 91.24
NEW BEST: epoch 9 valid F1 94.31 best test F1 90.99
NEW BEST: epoch 10 valid F1 94.73 best test F1 91.37
NEW BEST: epoch 11 valid F1 94.96 best test F1 92.3
NEW BEST: epoch 12 valid F1 95.56 best test F1 92.51
NEW BEST: epoch 13 valid F1 95.8 best test F1 92.44

NEW BEST: epoch 15 valid F1 96.11 best test F1 

In [6]:
import hw4a

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 3,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 90,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}

exp4 = hw4a.test_rnnslu(**param)

verbose: True
normal: True
win: 3
savemodel: False
fold: 3
seed: 345
emb_dimension: 90
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 70.43 best test F1 67.48
NEW BEST: epoch 1 valid F1 80.48 best test F1 76.76
NEW BEST: epoch 2 valid F1 82.68 best test F1 79.46
NEW BEST: epoch 3 valid F1 87.11 best test F1 82.67
NEW BEST: epoch 4 valid F1 88.94 best test F1 84.88
NEW BEST: epoch 5 valid F1 91.03 best test F1 86.51

NEW BEST: epoch 7 valid F1 91.92 best test F1 89.77
NEW BEST: epoch 8 valid F1 93.42 best test F1 90.76
NEW BEST: epoch 9 valid F1 94.04 best test F1 91.28

NEW BEST: epoch 11 valid F1 94.52 best test F1 91.94
NEW BEST: epoch 12 valid F1 94.83 best test F1 92.37
NEW BEST: epoch 13 valid F1 94.95 best test F1 91.9

NEW BEST: epoch 15 valid F1 95.83 best test F1 92.46

NEW BEST: epoch 17 valid F1 95.93 best test F1 93.66
NEW BEST: epoch 18 valid F1 95.98 best tes

###### I have experimented with [(win:7, emb_dimension:50), (win:11, emb_dimension:50), (win:7, emb_dimension:70), (win:3, emb_dimension:90)]. Best result I got was F1 score of 94.2 with (win:7, emb_dimension:50).

#### Bullet 2: Experiment with RNNSLU without normalizing word embeddings

###### From part I, best result came with running by default parameters. That is, when the size of context window is 7 and dimenions of the word embeddings is 50

In [ ]:
import hw4a

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}

res = hw4a.test_rnnslu(**param)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: False
win: 7
layer_norm: False
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.96 best test F1 63.35
NEW BEST: epoch 1 valid F1 81.28 best test F1 76.9
NEW BEST: epoch 2 valid F1 84.73 best test F1 81.46
NEW BEST: epoch 3 valid F1 88.04 best test F1 84.06
NEW BEST: epoch 4 valid F1 90.05 best test F1 86.02
NEW BEST: epoch 5 valid F1 91.31 best test F1 86.93
NEW BEST: epoch 6 valid F1 92.65 best test F1 88.75
NEW BEST: epoch 7 valid F1 93.21 best test F1 89.79
NEW BEST: epoch 8 valid F1 94.0 best test F1 90.6
NEW BEST: epoch 9 valid F1 94.86 best test F1 90.68
NEW BEST: epoch 10 valid F1 95.13 best test F1 91.98

NEW BEST: epoch 12 valid F1 95.68 best test F1 92.92
NEW BEST: epoch 13 valid F1 95.91 best test F1 92.79

NEW BEST: epoch 15 valid F1 96.25 best test F1 93.48
NEW BEST: epoch 16 valid F1 

##### Without normalizing the word embeddings after each update we got almost same, though a bit lower result than that of with normalizing.

#### Bullet 3: Implement RNN with two hidden layers

In [1]:
import hw4a
import imp
imp.reload(hw4a)

param = {'fold': 3,
        'lr': 0.0970806646812754,
        'verbose': True,
        'decay': True,
        'win': 7,
        'nhidden1': 200,
        'nhidden2': 100,
        'seed': 345,
        'emb_dimension': 50,
        'nepochs': 60,
        'savemodel': False,
        'normal': True,
        'folder':'../result'}

res = hw4a.test_rnnslu2(**param)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
nhidden2: 100
win: 7
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 57.63 best test F1 53.13
NEW BEST: epoch 1 valid F1 68.13 best test F1 66.74
NEW BEST: epoch 2 valid F1 75.54 best test F1 74.28
NEW BEST: epoch 3 valid F1 79.96 best test F1 77.21
NEW BEST: epoch 4 valid F1 84.54 best test F1 81.05
NEW BEST: epoch 5 valid F1 86.04 best test F1 82.64
NEW BEST: epoch 6 valid F1 86.94 best test F1 83.4
NEW BEST: epoch 7 valid F1 89.72 best test F1 85.89

NEW BEST: epoch 9 valid F1 91.22 best test F1 87.62
NEW BEST: epoch 10 valid F1 91.54 best test F1 88.46
NEW BEST: epoch 11 valid F1 92.35 best test F1 89.17
NEW BEST: epoch 12 valid F1 92.95 best test F1 90.12
NEW BEST: epoch 13 valid F1 94.06 best test F1 90.29

NEW BEST: epoch 15 valid F1 94.81 best test F1 91.27

NEW BEST: epoch 17 valid F1 

##### As we can see our accuracy did not imporve while using two layers. On the contrary it decreased a bit. Thus, in this case it would not make sense to increase number of layers to achive better accuracy.

#### Bullet 4: Experiment with layer normalization

In [6]:
import hw4a
import imp
imp.reload(hw4a)

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'layer_norm': True,
    'folder':'../result'}

res = hw4a.test_rnnslu(**param)

verbose: True
normal: False
win: 7
layer_norm: True
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 66.44 best test F1 60.9
NEW BEST: epoch 1 valid F1 81.89 best test F1 79.07
NEW BEST: epoch 2 valid F1 86.78 best test F1 83.95
NEW BEST: epoch 3 valid F1 91.32 best test F1 87.82
NEW BEST: epoch 4 valid F1 92.11 best test F1 88.13
NEW BEST: epoch 5 valid F1 93.67 best test F1 90.56

NEW BEST: epoch 7 valid F1 95.02 best test F1 92.48
NEW BEST: epoch 8 valid F1 95.33 best test F1 92.82
NEW BEST: epoch 9 valid F1 95.7 best test F1 93.36
NEW BEST: epoch 10 valid F1 95.79 best test F1 93.03
NEW BEST: epoch 11 valid F1 96.17 best test F1 93.48
NEW BEST: epoch 12 valid F1 96.47 best test F1 93.61
NEW BEST: epoch 13 valid F1 96.63 best test F1 93.48
NEW BEST: epoch 14 valid F1 96.73 best test F1 93.6


NEW BEST: epoch 17 valid F1

In [25]:
import hw4a
import imp
imp.reload(hw4a)

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'layer_norm': True,
    'folder':'../result'}

res = hw4a.test_rnnslu(**param)

verbose: True
normal: False
win: 11
layer_norm: True
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.36 best test F1 64.91
NEW BEST: epoch 1 valid F1 83.42 best test F1 79.91
NEW BEST: epoch 2 valid F1 87.79 best test F1 85.08
NEW BEST: epoch 3 valid F1 91.56 best test F1 87.62
NEW BEST: epoch 4 valid F1 92.83 best test F1 89.25
NEW BEST: epoch 5 valid F1 94.25 best test F1 90.84
NEW BEST: epoch 6 valid F1 94.42 best test F1 92.06
NEW BEST: epoch 7 valid F1 95.69 best test F1 93.06

NEW BEST: epoch 9 valid F1 95.98 best test F1 93.54
NEW BEST: epoch 10 valid F1 96.51 best test F1 93.28

NEW BEST: epoch 12 valid F1 96.63 best test F1 93.78


NEW BEST: epoch 15 valid F1 96.89 best test F1 93.75


NEW BEST: epoch 18 valid F1 96.9 best test F1 93.83


NEW BEST: epoch 21 valid F1 96.92 best test F1 94.16



NEW BEST: epoch 

In [1]:
import hw4a
import imp
imp.reload(hw4a)

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 70,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'layer_norm': True,
    'folder':'../result'}

res = hw4a.test_rnnslu(**param)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: False
win: 7
layer_norm: True
savemodel: False
fold: 3
seed: 345
emb_dimension: 70
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 72.81 best test F1 68.12
NEW BEST: epoch 1 valid F1 85.68 best test F1 80.9
NEW BEST: epoch 2 valid F1 89.59 best test F1 86.26
NEW BEST: epoch 3 valid F1 91.61 best test F1 88.68
NEW BEST: epoch 4 valid F1 93.13 best test F1 89.27
NEW BEST: epoch 5 valid F1 94.06 best test F1 90.56

NEW BEST: epoch 7 valid F1 95.42 best test F1 93.3
NEW BEST: epoch 8 valid F1 95.6 best test F1 93.1
NEW BEST: epoch 9 valid F1 95.82 best test F1 93.83
NEW BEST: epoch 10 valid F1 96.04 best test F1 93.67

NEW BEST: epoch 12 valid F1 96.38 best test F1 93.77
NEW BEST: epoch 13 valid F1 96.44 best test F1 93.77
NEW BEST: epoch 14 valid F1 96.61 best test F1 93.74
NEW BEST: epoch 15 valid F1 96.65 best test F1 93.77

NEW BEST: epoch 17 valid F1 

In [2]:
import hw4a
import imp
imp.reload(hw4a)

param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 3,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 90,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'layer_norm': True,
    'folder':'../result'}

res = hw4a.test_rnnslu(**param)

verbose: True
normal: False
win: 3
layer_norm: True
savemodel: False
fold: 3
seed: 345
emb_dimension: 90
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 76.36 best test F1 72.33
NEW BEST: epoch 1 valid F1 84.89 best test F1 81.66
NEW BEST: epoch 2 valid F1 89.25 best test F1 86.13
NEW BEST: epoch 3 valid F1 91.25 best test F1 88.23
NEW BEST: epoch 4 valid F1 92.01 best test F1 88.88
NEW BEST: epoch 5 valid F1 93.53 best test F1 90.15

NEW BEST: epoch 7 valid F1 94.78 best test F1 92.29
NEW BEST: epoch 8 valid F1 95.18 best test F1 93.05
NEW BEST: epoch 9 valid F1 95.62 best test F1 92.64

NEW BEST: epoch 11 valid F1 96.5 best test F1 93.54

NEW BEST: epoch 13 valid F1 96.55 best test F1 93.38

NEW BEST: epoch 15 valid F1 96.62 best test F1 93.62

NEW BEST: epoch 17 valid F1 96.84 best test F1 93.72
NEW BEST: epoch 18 valid F1 96.99 best test F1 93.34
NEW BEST: epoch 19 vali

##### In general we can se little improvment while using layer normalization. That is, all of the test results for the epxeriment achieved score of more than 94 while without layer normalization only one achieved this.

#### Bullet 5: Observation on embedding pairs

In [10]:
from hw4_utils import load_data
import numpy
train_set, valid_set, test_set, dic = load_data(3)
embeddings = numpy.load('../result/embeddings.npy')

In [11]:
dic['words2idx']

{"'d": 0,
 "'ll": 1,
 "'m": 2,
 "'re": 3,
 "'s": 4,
 "'t": 5,
 '72s': 6,
 '<UNK>': 7,
 'DIGIT': 8,
 'DIGITDIGIT': 9,
 'DIGITDIGITDIGIT': 10,
 'DIGITDIGITDIGITDIGIT': 11,
 'DIGITDIGITDIGITDIGITDIGITDIGIT': 12,
 'a': 13,
 'abbreviation': 14,
 'abbreviations': 15,
 'about': 16,
 'ac': 17,
 'actually': 18,
 'after': 19,
 'afternoon': 20,
 'again': 21,
 'air': 22,
 'aircraft': 23,
 'airfare': 24,
 'airline': 25,
 'airlines': 26,
 'airplane': 27,
 'airplanes': 28,
 'airport': 29,
 'airports': 30,
 'alaska': 31,
 'all': 32,
 'along': 33,
 'also': 34,
 'am': 35,
 'america': 36,
 'american': 37,
 'amount': 38,
 'an': 39,
 'and': 40,
 'angeles': 41,
 'another': 42,
 'any': 43,
 'anywhere': 44,
 'ap': 45,
 'ap57': 46,
 'ap80': 47,
 'approximately': 48,
 'april': 49,
 'are': 50,
 'area': 51,
 'arizona': 52,
 'around': 53,
 'arrange': 54,
 'arrangements': 55,
 'arrival': 56,
 'arrivals': 57,
 'arrive': 58,
 'arrives': 59,
 'arriving': 60,
 'as': 61,
 'at': 62,
 'atl': 63,
 'atlanta': 64,
 'august':

In [12]:
dic_words2idx = dic['words2idx']
words2vectors = dict()
for key, value in dic_words2idx.items():
    words2vectors[key] = embeddings[value]

In [13]:
words2vectors

{"'d": array([-0.0506615 ,  0.01984945, -0.00110418,  0.00772954, -0.28261951,
        -0.01718201,  0.09337538, -0.04040844, -0.16952023,  0.06161614,
         0.13995777, -0.2254795 , -0.14916934,  0.07119441,  0.31522855,
        -0.09523797, -0.10040957,  0.06341674, -0.23349342,  0.06729737,
         0.22828004, -0.07392453, -0.04541607, -0.01992875,  0.23328914,
         0.05870074, -0.00595046, -0.02929429, -0.02397818, -0.16776201,
        -0.01166161, -0.20912217,  0.10438107, -0.16034985,  0.33052391,
         0.21534778,  0.00225402, -0.16716219,  0.0663885 ,  0.11995836,
        -0.27752185,  0.12656529,  0.12789892, -0.11044709,  0.15532292,
         0.0362653 , -0.04786768, -0.10535812, -0.0367941 ,  0.16402622], dtype=float32),
 "'ll": array([ 0.01785998, -0.07101957,  0.11307583, -0.14474712, -0.14750449,
         0.19727679,  0.01479247,  0.005991  ,  0.0440961 ,  0.15833534,
        -0.19459321, -0.04915843, -0.00537258,  0.04866338, -0.14606324,
        -0.20997767, 

###### Findings are in the comments

In [14]:
from scipy.spatial.distance import cosine 

# I could observe several interesting patterns from the data
# First of all lets have a look if we could learn that we are looking for an airline to fly to Florida 
print(cosine(words2vectors['florida'] - words2vectors['delta'], words2vectors['florida'] - words2vectors['united']))
# prints 0.136188182621, which is low enough to confirm we could learn that we are looking for an airline for Florida

# What if I try Lufthansa instead of United from previos
print(cosine(words2vectors['florida'] - words2vectors['delta'], words2vectors['florida'] - words2vectors['lufthansa']))
# prints 0.410835894191. We see big increase between local airline and international airline.
# Perhaps people are more accustomed for local airlines, and look for local airlines for local flights.
# Thus, we had much smaller results for Delta vs United rather than Delta vs Lufthansa

# Lets try other local airlines to see if pattern persists
print(cosine(words2vectors['florida'] - words2vectors['delta'], words2vectors['florida'] - words2vectors['continental']))
# prints 0.119208534852
print(cosine(words2vectors['florida'] - words2vectors['delta'], words2vectors['florida'] - words2vectors['american']))
# prints 0.149490773751
# As we have expected Delta vs local airlines are much closer rather than Delta vs Lufthansa.
# This makes sense since, we would expect some kind of similar patterns for local airlines for local flights
# rather than international airlines for local flights

# Lets look at other results
print(cosine(words2vectors['florida'] - words2vectors['flight'], words2vectors['georgia'] - words2vectors['flight']))
# prints 0.079974350448
# We are learning that Florida and Georgia are states which you can take flights to

# Lets see if we are learning months
print(cosine(words2vectors['november'] - words2vectors['month'], words2vectors['december'] - words2vectors['month']))
# prints 0.210565320508
# Looks like we are indeed learning that November and December are months

# Lets see how breakfast and dinner related in terms of cost
print(cosine(words2vectors['dinner'] - words2vectors['cost'], words2vectors['breakfast'] - words2vectors['cost']))
# prints 0.179627447847
# We are getting some kind of relation between breakfast and dinner in terms of cost

# Lets see how cost and dollars are related
print(cosine(words2vectors['highest'] - words2vectors['cost'], words2vectors['lowest'] - words2vectors['cost']))
# prints 0.286330131367
print(cosine(words2vectors['highest'] - words2vectors['dollars'], words2vectors['lowest'] - words2vectors['dollars']))
# prints 0.230120793979
# First of all we see that highest and lowest are related in terms of cost.
# Also we are seeing that they are related to dollars with very close degree to that of in terms of cost

0.136188182621
0.410835894191
0.119208534852
0.149490773751
0.079974350448
0.210565320508
0.179627447847
0.286330131367
0.230120793979


### Problem b: Learning Parity Function
#### Bullet 1: Shallow MLP

In [93]:
import hw4b[]
import imp
imp.reload(hw4b)
hw4b.test_mlp_parity(n_bit = 8, n_hidden = 32, n_hiddenLayers = 1, learning_rate = 0.5, batch_size = 50, n_epochs = 500)

... building the model
epoch 1, minibatch 20/20, validation error 51.400000 %
     epoch 1, minibatch 20/20, test error of best model 47.000000 %
epoch 2, minibatch 20/20, validation error 51.200000 %
     epoch 2, minibatch 20/20, test error of best model 49.000000 %
epoch 3, minibatch 20/20, validation error 53.000000 %
epoch 4, minibatch 20/20, validation error 51.200000 %
epoch 5, minibatch 20/20, validation error 52.600000 %
epoch 6, minibatch 20/20, validation error 52.600000 %
epoch 7, minibatch 20/20, validation error 52.400000 %
epoch 8, minibatch 20/20, validation error 52.000000 %
epoch 9, minibatch 20/20, validation error 52.400000 %
epoch 10, minibatch 20/20, validation error 53.000000 %
epoch 11, minibatch 20/20, validation error 53.400000 %
epoch 12, minibatch 20/20, validation error 51.400000 %
epoch 13, minibatch 20/20, validation error 50.600000 %
     epoch 13, minibatch 20/20, test error of best model 51.000000 %
epoch 14, minibatch 20/20, validation error 50.600000

The training process for function test_mlp_parity ran for 0.12m


In [2]:
import hw4b
import imp
imp.reload(hw4b)

hw4b.test_mlp_parity(n_bit = 12, n_hidden = 128, n_hiddenLayers = 1, learning_rate = 0.2, batch_size = 10, n_epochs = 3000)

... building the model
epoch 1, minibatch 100/100, validation error 48.600000 %
     epoch 1, minibatch 100/100, test error of best model 49.000000 %
epoch 2, minibatch 100/100, validation error 49.000000 %
epoch 3, minibatch 100/100, validation error 48.200000 %
     epoch 3, minibatch 100/100, test error of best model 45.000000 %
epoch 4, minibatch 100/100, validation error 48.800000 %
epoch 5, minibatch 100/100, validation error 48.800000 %
epoch 6, minibatch 100/100, validation error 47.800000 %
     epoch 6, minibatch 100/100, test error of best model 41.000000 %
epoch 7, minibatch 100/100, validation error 48.200000 %
epoch 8, minibatch 100/100, validation error 48.600000 %
epoch 9, minibatch 100/100, validation error 49.400000 %
epoch 10, minibatch 100/100, validation error 48.600000 %
epoch 11, minibatch 100/100, validation error 48.000000 %
epoch 12, minibatch 100/100, validation error 47.400000 %
     epoch 12, minibatch 100/100, test error of best model 45.000000 %
epoch 13,

The training process for function test_mlp_parity ran for 2.90m


###### I could achieve 100% test accuracy for 8 bit case while using (n_hidden = 32, n_hiddenLayers = 1, learning_rate = 0.5, batch_size = 50).
###### For 12 bit case I could achieve 100% validation accuracy and 99% test accuracy while using (n_hidden = 128, n_hiddenLayers = 1, learning_rate = 0.2, batch_size = 10)

#### Bullet 2: Deep MLP

In [78]:
import hw4b
import imp
imp.reload(hw4b)
hw4b.test_mlp_parity(n_bit = 8, n_hidden = 8, n_hiddenLayers = 2, learning_rate = 0.1, batch_size = 50, n_epochs = 500)

... building the model
epoch 1, minibatch 20/20, validation error 54.200000 %
     epoch 1, minibatch 20/20, test error of best model 50.000000 %
epoch 2, minibatch 20/20, validation error 51.400000 %
     epoch 2, minibatch 20/20, test error of best model 49.000000 %
epoch 3, minibatch 20/20, validation error 50.400000 %
     epoch 3, minibatch 20/20, test error of best model 46.000000 %
epoch 4, minibatch 20/20, validation error 50.200000 %
     epoch 4, minibatch 20/20, test error of best model 45.000000 %
epoch 5, minibatch 20/20, validation error 51.400000 %
epoch 6, minibatch 20/20, validation error 52.000000 %
epoch 7, minibatch 20/20, validation error 53.000000 %
epoch 8, minibatch 20/20, validation error 53.000000 %
epoch 9, minibatch 20/20, validation error 54.000000 %
epoch 10, minibatch 20/20, validation error 55.200000 %
epoch 11, minibatch 20/20, validation error 55.000000 %
epoch 12, minibatch 20/20, validation error 55.000000 %
epoch 13, minibatch 20/20, validation erro

The training process for function test_mlp_parity ran for 0.15m


In [86]:
import hw4b
import imp
imp.reload(hw4b)
hw4b.test_mlp_parity(n_bit = 12, n_hidden = 12, n_hiddenLayers = 3, learning_rate = 0.1, batch_size = 50, n_epochs = 1000)

... building the model
epoch 1, minibatch 20/20, validation error 50.800000 %
     epoch 1, minibatch 20/20, test error of best model 46.000000 %
epoch 2, minibatch 20/20, validation error 50.400000 %
     epoch 2, minibatch 20/20, test error of best model 44.000000 %
epoch 3, minibatch 20/20, validation error 49.200000 %
     epoch 3, minibatch 20/20, test error of best model 46.000000 %
epoch 4, minibatch 20/20, validation error 49.800000 %
epoch 5, minibatch 20/20, validation error 50.600000 %
epoch 6, minibatch 20/20, validation error 50.800000 %
epoch 7, minibatch 20/20, validation error 50.800000 %
epoch 8, minibatch 20/20, validation error 50.000000 %
epoch 9, minibatch 20/20, validation error 50.200000 %
epoch 10, minibatch 20/20, validation error 51.000000 %
epoch 11, minibatch 20/20, validation error 51.400000 %
epoch 12, minibatch 20/20, validation error 50.400000 %
epoch 13, minibatch 20/20, validation error 49.800000 %
epoch 14, minibatch 20/20, validation error 49.600000 

The training process for function test_mlp_parity ran for 0.34m


##### I could achieve 100% test accuracy for 8 bit while using (n_hidden = 8, n_hiddenLayers = 2, learning_rate = 0.1, batch_size = 50)
##### I could achieve 100% test accuracy for 12 bit while using (n_hidden = 12, n_hiddenLayers = 3, learning_rate = 0.1, batch_size = 50)

#### Bullet 3: RNN

In [3]:
import hw4b
import imp
imp.reload(hw4b)
param = {
    'n_bit': 8,
    'lr': 0.9,
    'verbose': True,
    'decay': False,
    'nhidden': 4,
    'seed': 345,
    'nepochs': 100}
hw4b.test_rnn_parity(**param)

nhidden: 4
verbose: True
decay: False
n_bit: 8
nepochs: 100
seed: 345
lr: 0.9
... building the model
... training
('Epoch:', 0)
NEW BEST: epoch 0 valid 47.40 best test 61.00
('Epoch:', 1)
('Epoch:', 2)
NEW BEST: epoch 2 valid 100.00 best test 100.00
BEST RESULT: epoch 2 valid 100.00 best test 100.00


In [5]:
import hw4b
import imp
imp.reload(hw4b)
param = {
    'n_bit': 12,
    'lr': 0.9,
    'verbose': True,
    'decay': False,
    'nhidden': 6,
    'seed': 345,
    'nepochs': 100}
hw4b.test_rnn_parity(**param)

nhidden: 6
verbose: True
decay: False
n_bit: 12
nepochs: 100
seed: 345
lr: 0.9
... building the model
... training
('Epoch:', 0)
NEW BEST: epoch 0 valid 50.20 best test 50.00
('Epoch:', 1)
('Epoch:', 2)
('Epoch:', 3)
('Epoch:', 4)
('Epoch:', 5)
('Epoch:', 6)
('Epoch:', 7)
('Epoch:', 8)
('Epoch:', 9)
('Epoch:', 10)
('Epoch:', 11)
('Epoch:', 12)
('Epoch:', 13)
('Epoch:', 14)
('Epoch:', 15)
('Epoch:', 16)
('Epoch:', 17)
('Epoch:', 18)
NEW BEST: epoch 18 valid 50.40 best test 50.00
('Epoch:', 19)
('Epoch:', 20)
('Epoch:', 21)
('Epoch:', 22)
('Epoch:', 23)
('Epoch:', 24)
('Epoch:', 25)
('Epoch:', 26)
('Epoch:', 27)
('Epoch:', 28)
NEW BEST: epoch 28 valid 55.80 best test 50.00
('Epoch:', 29)
('Epoch:', 30)
('Epoch:', 31)
NEW BEST: epoch 31 valid 100.00 best test 100.00
BEST RESULT: epoch 31 valid 100.00 best test 100.00


##### I have achieved 100% for 8 bits and 12 bits cases using 4 and 6 hidden units, respectively. 

#### Bullet 4: LSTM

In [4]:
import hw4b
import imp
imp.reload(hw4b)
param = {
    'n_bit': 8,
    'lr': 1.2,
    'verbose': True,
    'decay': False,
    'nhidden': 1,
    'seed': 345,
    'nepochs': 200}
hw4b.test_lstm_parity(**param)

nhidden: 1
verbose: True
decay: False
n_bit: 8
nepochs: 200
seed: 345
lr: 1.2
... building the model
... training
('Epoch:', 0)
NEW BEST: epoch 0 valid 100.00 best test 100.00
BEST RESULT: epoch 0 valid 100.00 best test 100.00


In [3]:
import hw4b
import imp
imp.reload(hw4b)
param = {
    'n_bit': 12,
    'lr': 1.2,
    'verbose': True,
    'decay': False,
    'nhidden': 1,
    'seed': 345,
    'nepochs': 200}
hw4b.test_lstm_parity(**param)

nhidden: 1
verbose: True
decay: False
n_bit: 12
nepochs: 200
seed: 345
lr: 0.9
... building the model
... training
('Epoch:', 0)
NEW BEST: epoch 0 valid 50.20 best test 50.00
('Epoch:', 1)
NEW BEST: epoch 1 valid 100.00 best test 100.00
BEST RESULT: epoch 1 valid 100.00 best test 100.00


##### I have achieved 100% for 8 bits and 12 bits cases using only 1 hidden unit